File: graph/DFS/iterative/java/DFS.java
import java.util.*;

public class DFS {
    public static List<Integer> DFS(List<List<Integer>> graph, int start) {
        List<Integer> order = new ArrayList<>();
        boolean[] visited = new boolean[graph.size()];
        Deque<Integer> stack = new ArrayDeque<>();
        stack.push(start);

        while (!stack.isEmpty()) {
            int node = stack.pop();

            if (visited[node]) {
                continue;
            }

            visited[node] = true;
            order.add(node);

            List<Integer> neighbors = graph.get(node);
            for (int i = neighbors.size() - 1; i >= 0; i--) {
                int neighbor = neighbors.get(i);
                if (!visited[neighbor]) {
                    stack.push(neighbor);
                }
            }
        }

        return order;
    }
}

File: graph/DFS/iterative/cpp/DFS.cpp
#include "DFS.hpp"

#include <vector>

using namespace std;

vector<int> DFS(const vector<vector<int>>& graph, int start) {
    vector<int> order;
    vector<bool> visited(graph.size(), false);
    vector<int> stack = {start};

    while (!stack.empty()) {
        const int node = stack.back();
        stack.pop_back();

        if (visited[node]) {
            continue;
        }

        visited[node] = true;
        order.push_back(node);

        for (auto it = graph[node].rbegin(); it != graph[node].rend(); ++it) {
            if (!visited[*it]) {
                stack.push_back(*it);
            }
        }
    }

    return order;
}
